### DISCUSSION TITLE
[HSS 2022.03 with PMP enabled] HSS hands-up because "MMC->SRS12" keeps Zero (ms_mmc.c::mmc_main_plic_IRQHandler())

### DISCUSSION THREAD

**Discussion by okawaj9** [polarfire-soc/polarfire-soc-documentation#195] [HSS and U-Boot] [answered]
Hello,

When enabling PMP with HSS v2202.03, HSS hands up in MSS_MMC_sdma_read() and then it does not boot-up successfully. As long as I debugged,  it's caused because "MMC->SRS12" keeps Zero (ms_mmc.c::mmc_main_plic_IRQHandler()).

* "PMP enabled" MSS_Config setting is based on the published document ([mpfs-pmp-demo.md](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/demo-guides/mpfs-pmp-demo.md)).   
* To get successful HSS boot with enabling PMP, could you please advice what operation user needs to add with HSS v2022.03 ?

* (Following logs are experiments with "SDcard" as you can see below.  But I also confirmed that "eMMC" experiment (without SDcard) causes same phenomenon (hand-up))
 
 
 
**[1] Phenomenon of HSS 2022.03 hand-up when enabling PMP:**
* File: HSS_log_2022p03_PMP_enabled.png
![HSS_log_2022p03_PMP_enabled](https://user-images.githubusercontent.com/61904969/187117700-efd30508-f8eb-4997-91d0-f99e19c0a63a.png)

**[2] MSS Config settings and files:**
* MSS Config File (zip)
[ICICLE_MSS.zip](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9441352/ICICLE_MSS.zip)
 * MSS Config settings:
 Base cfg --> icicle-kit-reference-design-2022.03/script_support/MPFS_ICICLE_MSS_baremetal.cfg

 PMP Eanbled:
 * "Memory Partition and Protection" --> "Use Processor PMP and AXI Switch MPU Configurations"

 Disabling setting to prevent error on MSS Configurator:
 * "Memory Partition and Protection" --> "Processor PMP" -->
   GPIO_0, GPIO_1, User Crypro Configuration, QSPI, User Crypto Memory

 Add Memory Region...
 * "Memory Partition and Protection" --> "Processor PMP" --> "Add Memory Region..."
  DDR_Cached_1GB_0 : DDR_Cached_1GB : AMP_Context_A : 0x8000_0000 : 2_GB : 0x801F_FFFF : check all 
  DDR_Cached_1GB_1 : DDR_Cached_1GB : AMP_Context_B : 0x8020_0000 : 2_GB : 0x803F_FFFF : check all 

 (Screenshot) MSS_Config_PMP_enabled.png
![MSS_Config_PMP_enabled](https://user-images.githubusercontent.com/61904969/187118287-fc722421-b30e-44e5-b814-d58776442b69.png)

**[3] Copy PMP enabled MSS Config output (XML file) into HSS:**

* Copied the "ICICLE_MSS_mss_cfg.xml" into boards/mpfs-icicle-kit-es/soc_fpga_design/xml/.

* Removed files under "boards/mpfs-icicle-kit-es/fpga_design_config/" before build HSS.

**[4] HSS 2022.03 Debug Log:**

* Added debug messages in mmc_main_plic_IRQHandler()  (File: [mss_mmc.zip](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9441394/mss_mmc.zip))
   hss-software-services/
    baremetal/
     polarfire-soc-bare-metal-library/
      src/
       platform/
        drivers/
         mss/
          mss_mmc/
           mss_mmc.c

 (HSS_debug_log_2022p03_PMP_enabled.png)![HSS_debug_log_2022p03_PMP_enabled](https://user-images.githubusercontent.com/61904969/187118882-a87a10f1-ba99-4bdd-86e4-4b0d9f0aa387.png)

**[5] Reference of HSS debug long when PMP disabled:**

 (HSS_debug_log_2022p3_PMP_disabled_0.png)![HSS_debug_log_2022p3_PMP_disabled_0](https://user-images.githubusercontent.com/61904969/187118720-4656d919-2170-49a6-bec0-e80ec3e5ff0e.png)
 (HSS_debug_log_2022p3_PMP_disabled_1.png)![HSS_debug_log_2022p3_PMP_disabled_1](https://user-images.githubusercontent.com/61904969/187118756-69915596-5add-4b26-a8e3-080b2d476308.png)

**[6] Another information to replicate this issue:**

 To get build "pass" with above "mms_mmc.c" including debug messages, added "#ifndef MIN" right before "#define MIN" in following files, like below

   \#ifndef MIN
   \#  define MIN(A,B)		((A) < (B) ? A : B)

   (a) 
   hss-software-services/
    baremetal/
     polarfire-soc-bare-metal-library/
      src/
       platform/
        mpfs_hal/
         common/
          bits.h

   (b)
   hss-software-services/
    include/
     hss_types.h

 
 

**[7] Another information 2:**
* I tried to conduct "import MSS" by using PMP enabled MSS_Config (above) into ICICLE-KIT Reference design 2022.03.  But this issue was not resolved (it keeps same phenomenon).

Thank you for your advice.

Best Regards,

**Comment by padmaraob** [comment]
I think, the MPU is not configured for the DDR(0x80000000) and LIM area(0x08000000) in the design for
the MMC DMA access. The MMC DMA works only when the MPU PMP regions
are configured with right memory region and size.

You can add below code in the HSS and try it.

https://github.com/polarfire-soc/hart-software-services/blob/master/services/mmc/mmc_api.c#L226
 
/* DMA init for MMC */

/* LIM memory with 2MB */
MSS_MPU_configure(MSS_MPU_MMC,
                                     MSS_MPU_PMP_REGION3,
                                     0x08000000,
                                     0x200000,
                                     MPU_MODE_READ_ACCESS|MPU_MODE_WRITE_ACCESS|MPU_MODE_EXEC_ACCESS,
                                    MSS_MPU_AM_NAPOT,
                                    0u);

/* DDR memory with 1GB */
MSS_MPU_configure(MSS_MPU_MMC,
                                     MSS_MPU_PMP_REGION1,
                                     0x80000000,
                                     0x40000000,
                                     MPU_MODE_READ_ACCESS|MPU_MODE_WRITE_ACCESS|MPU_MODE_EXEC_ACCESS,
                                    MSS_MPU_AM_NAPOT,
                                    0u);

**Reply by okawaj9** [reply]
Thank you for the suggestion.
But unfortunately, the suggestion did not resolve this problem ("MMC->SRS12" keeps Zero).  Attached the result.

![HSS_PMP_en_trial1](https://user-images.githubusercontent.com/61904969/189582966-5f5a37c4-4917-45d8-9535-52d317e886a8.png)

(At that time, added #include "mss_mpu.h" at line 39 of mmc_api.c to avoid build error)

Best Regards,

**Comment by padmaraob** [comment]
Can you please share the values at memory "0x20005700 " to "0x20005718 "?

You can capture these values using "Softconsole Memory view"

**Reply by okawaj9** [reply]
Attached.
![HSS_PMP_en_debug2](https://user-images.githubusercontent.com/61904969/189630981-d4919e47-c79c-458f-85a1-85d22e54b913.png)

**Comment by padmaraob** [comment, marked as answer]
the HSS is executing from Scratchpad not LIM so can you change MPU MMC configuration for Scratch pad.  

/* DMA init for MMC */

/* Scratchpad memory with 2MB */
MSS_MPU_configure(MSS_MPU_MMC,
MSS_MPU_PMP_REGION3,
0x0A000000,
0x200000,
MPU_MODE_READ_ACCESS|MPU_MODE_WRITE_ACCESS|MPU_MODE_EXEC_ACCESS,
MSS_MPU_AM_NAPOT,
0u);

/* DDR memory with 1GB */
MSS_MPU_configure(MSS_MPU_MMC,
MSS_MPU_PMP_REGION1,
0x80000000,
0x40000000,
MPU_MODE_READ_ACCESS|MPU_MODE_WRITE_ACCESS|MPU_MODE_EXEC_ACCESS,
MSS_MPU_AM_NAPOT,
0u);

**Reply by okawaj9** [reply]
[1] Issue of "MMC->SRS12 keeps Zero" seems resolved by the solution.  Thanks.
[2] On the other hand, U54_1 could not start application as  you can see attachment.  

Actually, this is a simple trial of "PMP enabling demo" without using YMODEM operation on PuTTY, **as real usage**.  Reference is https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/demo-guides/mpfs-pmp-demo.md, and using HSS v2022.03 (and ICICLE-KIT reference design v2022.3).  

So, if you already succeeded "enabling PMP and application run" without the YMODEM operation, could you please share with us what operation user needs to add in addition to https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/demo-guides/mpfs-pmp-demo.md ?

Thank you for your cooperation.

Best Regards,

![HSS_PMP_en_debug3_SpadSetting](https://user-images.githubusercontent.com/61904969/190318327-179faa12-8eae-413b-b4ca-6125a715378e.png)
